openwisp-config/files/openwisp.agent

```diff
@@ -133,11 +140,22 @@ register() {
 	local model=$(cat /tmp/sysinfo/model)
 	local os=$(ubus call system board | grep description | awk -F\" '{ print $4 }')
 	local system=$(ubus call system board | grep system | awk -F\" '{ print $4 }')
+	# get hardware id if script is given
+	if [ -x "$HARDWARE_ID_SCRIPT" ]; then
+		local hardware_id=$("$HARDWARE_ID_SCRIPT")
+	else
+		local hardware_id=""
+	fi
 	# prepare POST params
 	local params="backend=$backend"
-	# generate key from macaddress + shared secret
 	if [ "$CONSISTENT_KEY" == "1" ]; then
-		local consistent_key=$(echo -n "$macaddr+$SHARED_SECRET" | md5sum | awk '{print $1}')
+		if [ -x "$HARDWARE_ID_SCRIPT" ] && [ "$HARDWARE_ID_KEY" = "1" ]; then
+			# generate key from hardware id
+			local consistent_key=$(echo -n "$hardware_id+$SHARED_SECRET" | md5sum | awk '{print $1}')
+		else
+			# generate key from macaddress + shared secret
+			local consistent_key=$(echo -n "$macaddr+$SHARED_SECRET" | md5sum | awk '{print $1}')
+		fi
```

Comment:
the content of the if are mostly identical, maybe you can do something like:

```bash
if [ -x "$HARDWARE_ID_SCRIPT" ] && [ "$HARDWARE_ID_KEY" = "1" ]; then
    local base = $hardware_id
else
    local base = $macaddr
fi
local consistent_key=$(echo -n "$base+$SHARED_SECRET" | md5sum | awk '{print $1}')
```

Reply:
@nemesisdesign I added that change :-)